src/lib.rs

```diff
@@ -49,49 +49,57 @@ pub mod prelude {
 
 #[derive(Debug)]
 /// General return type of the Peel traversals
-pub struct PeelResult<'a> {
+pub struct PeelResult<'a, ParserId> {
     /// A vector of parser results
     pub result: ParserResultVec,
 
+    /// Identifiers of the parsers that matched
+    pub parsers: Vec<ParserId>,
+
     /// The left input
     pub left_input: &'a [u8],
 
     /// Possible error which occured during the parsing
     pub error: Option<PeelError>,
 }
 
-impl<'a> PeelResult<'a> {
+impl<'a, ParserId> PeelResult<'a, ParserId> {
     /// Create a new `ParserResult`
-    fn new(result: ParserResultVec, left_input: &'a [u8], error: Option<PeelError>) -> Self {
+    fn new(result: ParserResultVec, parsers: Vec<ParserId>, left_input: &'a [u8], error: Option<PeelError>) -> Self {
         PeelResult {
-            result: result,
-            left_input: left_input,
-            error: error,
+            result,
+            parsers,
+            left_input,
+            error,
         }
     }
 }
 
 /// The main peeling structure
-pub struct Peel<D> {
+pub struct Peel<D, ParserId = ()> {
     /// The memory arena of the tree
     pub graph: StableGraph<Parser<D>, ()>,
 
     /// The first node added will be the root
     pub root: Option<NodeIndex>,
 
+    /// Provides the parser IDs/hints.
+    id_map: HashMap<NodeIndex, ParserId>,
```

Comment:
Do you think it would be possible to use a faster data type than a hash map here?